File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.cpp
class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extra) {
        priority_queue<pair<double,int>> pq;
        int n = classes.size();
        int count=0;
        for(int i=0;i<n;i++)
        {
            int total = classes[i][1];
            int pass = classes[i][0];
            if(total == pass)
            {
                continue;
            }
            double avg = pass*1.0/total;
            pass++;
            total++;
            double nexAvg = pass*1.0/total;
            double change = nexAvg - avg;
            pq.push({change,i});
        }
        for(int i=1;i<=extra && !pq.empty();i++)
        {
            pair<double,int> p = pq.top();
            pq.pop();
            if(p.first ==0.0)
            {
                break;
            }
            classes[p.second][0]+=1;
            classes[p.second][1]+=1;
            int total = classes[p.second][1];
            int pass = classes[p.second][0];
            double avg = pass*1.0/total;
            pass++;
            total++;
            double nexAvg = pass*1.0/total;
            double change = nexAvg - avg;
            pq.push({change,p.second});
            // cout<<p.second<<" ";
        }
        double sum =0.0;
        for(int i=0;i<n;i++)
        {
            int total = classes[i][1];
            int pass = classes[i][0];
            double avg = pass*1.0/total;
            sum+=avg;
        }
        double average = sum/n;
        average = round(average * 1e5) / 1e5;
        return average;
    }
    
};

File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.java
class Solution {
    public double maxAverageRatio(int[][] classes, int extra) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            double gainA = (double) (a[1] - a[0]) / ((long) a[1] * (a[1] + 1));
            double gainB = (double) (b[1] - b[0]) / ((long) b[1] * (b[1] + 1));
            return Double.compare(gainB, gainA);
        });

        for (int[] c : classes) {
            pq.offer(new int[] { c[0], c[1] });
        }

        for (int i = 0; i < extra; i++) {
            int[] top = pq.poll();
            top[0] += 1;
            top[1] += 1;
            pq.offer(top);
        }
        
        double sum = 0.0;
        for (int[] c : pq) {
            sum += (double) c[0] / c[1];
        }

        return sum / classes.length;
    }
}